### PR TITLE
Hono command line interface

### PR DESCRIPTION
**Pull request by kaniyan**
Hono-cli (Command Line Interface) is now available under folder **cli**. A Hono Client can be started on a command line that logs the messages (telemetry, event or both) to the console. A command line argument named **message.type** tells the client what type of messages are to be received and logged.

### CODE REVIEW

**Comment by sophokles73** [comment]
this PR should not include the changes made in #718

File: cli/pom.xml [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        <artifactId>hono-bom</artifactId>
+        <version>0.7-M2-SNAPSHOT</version>
+        <relativePath>../bom</relativePath>
+    </parent>
+
+    <artifactId>hono-cli</artifactId>
+    <name>Hono command-line interface</name>
+    <url>https://www.eclipse.org/hono</url>
+
+    <description>Hone Command line Interface</description>
```
Comment (sophokles73) [inline cli/pom.xml]: It's called *Hono* ;-)

File: cli/pom.xml [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+    <properties>
+        <docker.skip.build>true</docker.skip.build>
+        <hono.messaging.host>hono-service-messaging.hono</hono.messaging.host>
+        <hono.registration.host>hono-service-device-registry.hono</hono.registration.host>
+        <hono.auth.host>hono-service-auth.hono</hono.auth.host>
+        <!--
+          The shared secret to use for creating/validating registration assertions.
+         -->
+        <hono.regAssertion.sharedSecret>g#aWO!BUm7aj*#%X*VGXKFhxkhNrMNj0</hono.regAssertion.sharedSecret>
```
Comment (sophokles73) [inline cli/pom.xml]: none of these properties except for `docker.skip.build` are relevant to the CLI module. Please remove ...

File: cli/src/main/java/org/eclipse/hono/cli/AbstractClient.java [hunk 3]
```diff
@@ -30,21 +30,24 @@
...
  */
-abstract class AbstractExampleClient {
+abstract class AbstractClient {
 
     protected final Logger LOG = LoggerFactory.getLogger(getClass());
 
     protected Context ctx;
     @Value(value = "${tenant.id}")
     protected String tenantId;
+    @Value(value = "${message.type:all}")
```
Comment (sophokles73) [inline cli/src/main/java/org/eclipse/hono/cli/AbstractClient.java]: as with the `tenant.id`, I'd rather have it set to the default value using the application.yml fiel (in which you already seem to set the default value).

File: cli/src/main/java/org/eclipse/hono/cli/Receiver.java [hunk 4]
```diff
@@ -10,80 +10,90 @@
...
 import io.vertx.core.Future;
 import io.vertx.proton.ProtonConnection;
 
+import java.util.ArrayList;
+import java.util.List;
+
 /**
- * Example of a event/telemetry receiver that connects to the Hono Server, waits for incoming messages and logs the message
- * payload if anything is received.
+ * Receiver that connects to the Hono Server, waits for incoming messages and logs the message payload if anything is
```
Comment (sophokles73) [inline cli/src/main/java/org/eclipse/hono/cli/Receiver.java]: In fact, there hasn't been a *Hono Server* component anymore for quite some time now. Let's call it simply *Hono* ...

File: cli/src/main/resources/logback-spring.xml [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        <appender-ref ref="STDOUT" />
+    </root>
+
+    <springProfile name="trace">
+        <logger name="org.eclipse.hono.cli" level="TRACE"/>
+        <logger name="org.eclipse.hono.client" level="TRACE"/>
+    </springProfile>
+
+    <springProfile name="dev">
+        <logger name="org.eclipse.hono.cli" level="TRACE"/>
```
Comment (sophokles73) [inline cli/src/main/resources/logback-spring.xml]: shouldn't that be `DEBUG`?

File: cli/src/test/java/org/eclipse/hono/cli/ReceiverTest.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+    /**
+     * Setups the receiver with mocks.
+     *
+     */
+    @Before
+    public void setup() {
+        final HonoClient client = mock(HonoClient.class);
+        receiver = new Receiver();
+        MockitoAnnotations.initMocks(this);
```
Comment (sophokles73) [inline cli/src/test/java/org/eclipse/hono/cli/ReceiverTest.java]: what is this good for?

File: cli/src/test/java/org/eclipse/hono/cli/ReceiverTest.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+     * @param context The vert.x test context.
+     */
+    @Test
+    public void testTelemetryStart(final TestContext context) {
+        receiver.messageType = "telemetry";
+        final Future<Void> done = Future.future();
+        done.setHandler(context.asyncAssertSuccess());
+        receiver.start().setHandler(result -> {
+            context.assertNotNull(result.result().list());
+            context.assertTrue(result.result().list().size() == 1);
```
Comment (sophokles73) [inline cli/src/test/java/org/eclipse/hono/cli/ReceiverTest.java]: you seem to assume that the AsyncResult is succeeded. You can actually make this explicit in the test by using:
```
receiver.start().setHandler(context.asyncAssertSuccess(result -> {
       context.assertNotNull(result.list());
       context.assertTrue(result.list().size() == 1);
}
```

File: cli/src/test/java/org/eclipse/hono/cli/ReceiverTest.java [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+     * 
+     * @param context The vert.x test context.
+     */
+    @Test
+    public void testEventStart(final TestContext context) {
+        receiver.messageType = "event";
+        final Future<Void> done = Future.future();
+        done.setHandler(context.asyncAssertSuccess());
+        receiver.start().setHandler(result -> {
+            context.assertNotNull(result.result().list());
```
Comment (sophokles73) [inline cli/src/test/java/org/eclipse/hono/cli/ReceiverTest.java]: see above

File: cli/src/test/java/org/eclipse/hono/cli/ReceiverTest.java [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+     * Verifies that the receiver is started successfully with message.type=all.
+     * 
+     * @param context The vert.x test context.
+     */
+    @Test
+    public void testDefaultStart(final TestContext context) {
+        final Future<Void> done = Future.future();
+        done.setHandler(context.asyncAssertSuccess());
+        receiver.start().setHandler(result -> {
+            context.assertNotNull(result.result().list());
```
Comment (sophokles73) [inline cli/src/test/java/org/eclipse/hono/cli/ReceiverTest.java]: see above

File: site/content/release-notes.md [hunk 10]
```diff
@@ -16,6 +16,7 @@ title = "Release Notes"
 
 * The Device Registry's AMQP endpoints can now be configured with the number of credits they should flow to clients connecting to the endpoints. The default value is 100. See [Device Registry admin guide]({{< relref "admin-guide/device-registry-config.md#service-configuration" >}}) for details.
 * Hono's service components can now be configured to use OpenSSL instead of the JVM's default SSL engine. The [admin guide]({{< relref "admin-guide/secure_communication.md#using-openssl" >}}) describes how to do this.
+* Hono-cli (Command Line Interface) is now available under folder `cli`. A command line argument `message.type` with value `telemetry`, `event` or `all` (default) tells the client what kind of messages to be received. See [Starting a Consumer]({{< relref "getting-started.md#starting-a-consumer" >}}) for more information.
```
Comment (sophokles73) [inline site/content/release-notes.md]: this should be moved to 0.7 section

**Comment by sophokles73** [comment]
@kaniyan can you also please rebase the PR?

**Comment by kaniyan** [comment]
@sophokles73 I made changes and rebased with PR.

**Pull request merged**